File: src/entities/BaseEntity.java
package entities;

import javax.xml.bind.annotation.XmlElement;
import javax.xml.bind.annotation.XmlRootElement;
import java.io.Serializable;
import java.util.UUID;

/**
 * Created by g.zubenko on 23.01.2017.
 */
@XmlRootElement
public abstract class BaseEntity implements Serializable {
    private long id;
    abstract String getView();

    public BaseEntity() {
        this.id = getNewId();
    }

    private long getNewId()
    {
        return -(UUID.randomUUID().getLeastSignificantBits());
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        BaseEntity that = (BaseEntity) o;

        return id == that.id;
    }

    @Override
    public int hashCode() {
        return (int) (id ^ (id >>> 32));
    }

    public long getId() {
        return id;
    }

    @XmlElement
    private void setId(long id){
        this.id=id;
    }
}

File: src/entities/BaseEntity.kt
package entities

import java.io.Serializable
import java.util.UUID
import javax.xml.bind.annotation.XmlElement
import javax.xml.bind.annotation.XmlRootElement

@XmlRootElement
abstract class BaseEntity : Serializable {
    var id: Long = -UUID.randomUUID().leastSignificantBits
        @XmlElement
        private set

    abstract val view: String

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || other !is BaseEntity) return false

        return id == other.id
    }

    override fun hashCode(): Int {
        return (id xor id.ushr(32)).toInt()
    }
}